profile/src/main/java/rdx/works/profile/domain/backup/RestoreProfileFromBackupUseCase.kt

```diff
@@ -1,83 +1,37 @@
-    private val mnemonicRepository: MnemonicRepository,
-    private val driveClient: DriveClient,
-    private val checkMigrationToNewBackupSystemUseCase: CheckMigrationToNewBackupSystemUseCase
+    private val checkMigrationToNewBackupSystemUseCase: CheckMigrationToNewBackupSystemUseCase,
+    private val sargonOsManager: SargonOsManager
 ) {
 
     suspend operator fun invoke(
         backupType: BackupType,
         mainSeedPhraseSkipped: Boolean
     ): Result<Unit> {
+        val sargonOs = sargonOsManager.sargonOs
+
         // always restore backup on mainnet
         val profile = backupProfileRepository.getTemporaryRestoringProfile(backupType)
             ?.changeGatewayToNetworkId(NetworkId.MAINNET) ?: return Result.failure(RuntimeException("No restoring profile available"))
 
-        val hostId = hostInfoRepository.getHostId()
-        val hostInfo = hostInfoRepository.getHostInfo()
-        val profileWithRestoredHeader = profile.claim(deviceInfo = DeviceInfo.from(hostId, hostInfo))
-
-        return if (mainSeedPhraseSkipped) {
-            mnemonicRepository.createNew()
-                .mapCatching { mnemonic ->
-                    val deviceFactorSource = FactorSource.Device.babylon(
-                        mnemonicWithPassphrase = mnemonic,
-                        hostInfo = hostInfo,
-                        createdAt = TimestampGenerator(),
-                        isMain = true
-                    )
-                    profileWithRestoredHeader.addMainBabylonDeviceFactorSource(mainBabylonFactorSource = deviceFactorSource)
-                }
-                .mapError { ProfileException.SecureStorageAccess }
-        } else {
-            Result.success(profileWithRestoredHeader)
-        }.then { profileToSave ->
-            if (backupType is BackupType.Cloud) {
-                Timber.tag("CloudBackup").d("☁\uFE0F Claiming Profile...")
-                driveClient.claimCloudBackup(
-                    file = backupType.entity,
-                    claimingProfile = profileToSave
-                ).onSuccess {
-                    Timber.tag("CloudBackup").d("☁\uFE0F Profile claimed, and now save it")
-                    profileRepository.saveProfile(profileToSave)
-                    backupProfileRepository.discardTemporaryRestoringSnapshot(backupType)
-                    checkMigrationToNewBackupSystemUseCase.revokeAccessToDeprecatedCloudBackup()
-                }.toUnitResult()
-            } else {
-                if (backupType is BackupType.DeprecatedCloud) {
-                    checkMigrationToNewBackupSystemUseCase.revokeAccessToDeprecatedCloudBackup()
-                }
-
-                Timber.tag("CloudBackup").d("Save profile")
-                profileRepository.saveProfile(profileToSave)
-                backupProfileRepository.discardTemporaryRestoringSnapshot(backupType)
-                Result.success(Unit)
+        return runCatching {
+            sargonOs.importWallet(
```

Comment:
These specific functions that mutate the profile, I am not sure are safe to be called in a different thread. The reason for that is how profile is stored [here](https://github.com/radixdlt/sargon/blob/752616f8bf26fdc071c78c835f5e3976ddc465aa/crates/sargon/src/system/sargon_os/profile_state_holder.rs#L24). This is an [`RWLock`](https://doc.rust-lang.org/stable/std/sync/struct.RwLock.html) which I am not sure it is thread safe.

Reply:
Unfortunately this lock is an `RWLock`, and the way sargon is developed, if such lock blocks the thread it panics and it does not "suspend". 

@GhenadieVP is aware of this, and this will be one of the improvements for `ProfileStateHolder` in the future.